### DISCUSSION TITLE
为什么开了"dnsLog"但看不到dns查询日志?

### DISCUSSION THREAD

**Discussion by GitCourser** [XTLS/Xray-core#3666] [Q&A]
"loglevel": "debug",
"dnsLog": true

如上, 把日志等级设成显示最多信息的 debug, 也看不到dns日志
access.log 文件里只有路由信息
error.log 文件里只有开头有创建 dns 的信息
2024/08/10 10:21:36 [Info] app/dns: DNS: created Remote DOH client for https://223.5.5.5/dns-query
2024/08/10 10:21:36 [Info] app/dns: DNS: created Remote DOH client for https://8.8.8.8/dns-query
但后面一直没出现dns查询

使用环境
Xray-core v1.8.23
Windows 10
chrome 用 SwitchyOmega 连接本机 xray 的 socks 代理端口
浏览器中没有启用"安全dns", 也没有启用"预加载网页"

请问要如何才能显示 dns 查询日志

**Comment by TrajanBenn** [comment]
domainStrategy设置成`IPIfNonMatch`或者`IPOnDemand`，不确定是否存在DNS泄漏

**Reply by GitCourser** [reply]
我是用的`AsIs`, 原来是要用`IPIfNonMatch`和`IPOnDemand`才能显示dns查询日志?

**Reply by TrajanBenn** [reply]
这种情况和Xray的路由策略有关

使用AsIs策略，DNS解析由代理完成，具体链路为

1. 浏览器访问google.com
2. Xray通过AsIs路由匹配发现这是一个外网地址
3. Xray将域名请求google.com发送给代理服务器
4. 代理服务器负责问询VPS的DNS服务器解析google.com的地址为142.250.69.206
5. 代理服务器发送请求到142.250.69.206
6. 数据返回Xray

使用IPOnDemand策略，DNS解析由本机完成，例如你的DNS模块设置为DOH，并使用<https://8.8.8.8/dns-query>查询外网地址，具体链路为

1. 浏览器访问google.com
2. Xray查询Xray自己的DNS缓存
3. 缓存不命中
4. Xray将google.com导入DNS模块
5. Xray路由匹配发现这是一个外网地址
6. Xray使用<https://8.8.8.8/dns-query>查询外网地址
7. Xray路由匹配发现8.8.8.8是国外IP
8. Xray将向<https://8.8.8.8/dns-query>的DNS查询请求发送给代理服务器
9. 代理服务器向<https://8.8.8.8/dns-query>发送请求获得<google.com>的IP是142.250.69.206
10. 数据返回
11. Xray记录缓存
12. Xray向142.250.69.206发送请求
13. Xray路由匹配发现142.250.69.206是一个外网地址
14. Xray将请求发送给代理服务器
15. 代理服务器发送请求到142.250.69.206
16. 数据返回Xray

实际观察IPIfNonMatch和IPOnDemand策略表现相似，不知道是不是系统设计的问题，项目源码看的不是很懂

也就是说AsIs策略DNS解析由代理服务器完成，IPOnDemandD策略DNS解析由本机客户端和代理服务器共同完成

~~但是实际通过WireShark抓包分析发现，使用AsIs策略并不能按照理想情况工作，多数情况下域名解析会由本机系统配置的DNS服务器解析(UDP明文模式向DNS服务器的53端口发送请请求)，造成DNS泄漏问题~~

~~采取IPOnDemand策略可以有效避免DNS泄漏(前提是合理配置路由和DNS解析对象)~~[#3702]已解决

AsIs在合理配置DNS解析策略的情况下不会有DNS查询日志

**Reply by GitCourser** [reply]
原来如此,感谢
